### ISSUE TITLE
Hint Resolve registers the same hint for a definitional class multiple times at different patterns

### ISSUE THREAD

**Issue by Janno** [coq/coq#19681]
### Description of the problem

`Hint Resolve` is a little overeager and finds multiple possible patterns for hints whose type ends in definitional classes. `Existing Instance` does not have this problem. The same issue exists with hints of non-class definitional types and it is independent of the opacity status of the class/type in the database. 

I think `Hint Resolve` is the only way that existing definitions can be registered into a specific database so this bug currently does not seem to have a workaround unless the database in question is `typeclass_instances` for which `Existing Instance` can be used.

AFAICT the documentation of `Hint Resolve` does not mention this odd behavior.

### Small Rocq / Coq file to reproduce the bug

```coq
Class Inj {A B} (f : B -> A) : Prop := inj : forall x y, f x = f y -> x = y.
Create HintDb test discriminated.
Hint Extern 100 (tt = tt) => reflexivity : test.
Axiom f : unit -> unit.
Axiom f_Inj : Inj f.
Module EI.
  #[local] Existing Instance f_Inj.      (* This registers [f_Inj] only with pattern [Inj f] *)
  Goal tt = tt. typeclasses eauto with test. Qed.
End EI.
Module HR.
  Hint Resolve f_Inj : test.    (* This adds an additional entry for [f_Inj] with pattern [_ = _] *)
  Goal tt = tt.
    Timeout 1 typeclasses eauto with test. (* The additional hint loops *)
  Abort.
End HR.
```

### Version of Rocq / Coq where this bug occurs

8.16-master

**Comment by Janno** [comment]
There might actually be one way to work around this issue using `#[local] Opaque Inj` just before `Hint Resolve` and undoing the effect later. It seems to work on the two Coq version I tested it on, 8.16 and 8.19.

**Comment by Janno** [comment]
The behavior is due to `make_resolves` allowing both `make_exact_entry` and `make_apply_entry` to succeed. They are only mutually exclusive if `hnf` is `false`. But for `Hint Resolve`, `hnf` is set to `true`. It seems that this was historically the case even for typeclass instances and @mattam82 fixed this for `Instance` in 4a1077f9d1ee00632ec72a4089013194f558cacd.

I wonder why the change was only made for typeclasses. It seems really counterintuitive not just for classes but for any kind of hint.

**Comment by ppedrot** [comment]
> I wonder why the change was only made for typeclasses

Probably backwards compatibility. If we can get rid of hnf for these commands it would be great imho.

**Comment by Janno** [comment]
A small test run reveals that this would not be an easy change. I can't even get through Bool.v because `not _`  is required to be unfolded in the conclusion of hints in order to trigger on a goal of type `False`. I thought maybe that could be fixed by specifying `False` as the explicit pattern but of course that is ignored for selecting the head symbol that the hint belongs to. This is going to be quite the rabbit hole and I don't think I have the time to go there.

**Comment by ppedrot** [comment]
> this would not be an easy change

You have your answer then...

**Comment by Janno** [comment]
So what is the way forward in this case? Add an `hnf` attribute to `Hint Resolve` to explicitly get the old behavior, warn every time `Hint Resolve` yields two hints when the attribute is not set, and eventually turn the warning into an error? The standard library can be ported wholesale to the new attribute and the same is true for every single project in coq ci that ens up failing when the warning is turned into an error.

**Comment by ppedrot** [comment]
Indeed, a warning when hnf is not a no-op + an attribute to force hnf and remove the warning would be a good solution.

**Comment by ppedrot** [comment]
Note also that IIRC we also call hnf on local hypotheses from the context when calling auto tactics and this can be expensive.
